File: file-handling/handling-csv-files/java/csv_file_append_optimised.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science

Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

class HandlingCSVFiles
{
    // The main method is the entry point for all Java programs
    public static void main(String[] args)
    {
        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter the title of the film: ");
        String title = scanner.nextLine();

        System.out.println("Enter the year the film was made: ");
        String year = scanner.nextLine();

        System.out.println("Enter the length of the film in minutes: ");
        String duration = scanner.nextLine();

        System.out.println("Enter the name of the film director: ");
        String director = scanner.nextLine();
        
        System.out.println("Enter the average film rating: ");
        String rating = scanner.nextLine();

        System.out.println("Enter the takings in US$ millions: ");
        String takings = scanner.nextLine();

        appendMovieOptimised(title, year, duration, director, rating, takings);
    }


    // Optimised version to append a new record to the CSV file
    public static void appendMovieOptimised(String title, String year, String duration, String director, String rating, String takings)
    {
        String newMovie = title + "," + year + "," + duration + "," + director + "," + rating + "," + takings;

        // Uses try-with-resources to automatically close the writers after execution
        try (FileWriter fw = new FileWriter("movies.csv", true);
             BufferedWriter writer = new BufferedWriter(fw)) {

            writer.write(newMovie);
            writer.newLine();
        } // The writers are now closed
        catch (IOException e) {
            e.printStackTrace();
        }
    }


}

File: file-handling/handling-csv-files/visual-basic/csv_file_append_optimised.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program

Module Program

    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        Console.WriteLine("Enter the title of the film: ")
        Dim title As String = Console.ReadLine()
        Console.WriteLine("Enter the year the film was made: ")
        Dim year As String = Console.ReadLine()
        Console.WriteLine("Enter the length of the film in minutes: ")
        Dim duration As String = Console.ReadLine()
        Console.WriteLine("Enter the name of the film director: ")
        Dim director As String = Console.ReadLine()
        Console.WriteLine("Enter the average film rating: ")
        Dim rating As String = Console.ReadLine()
        Console.WriteLine("Enter the takings in US$ millions: ")
        Dim takings As String = Console.ReadLine()
        AppendMovieOptimised(title, year, duration, director, rating, takings)
    End Sub

    Sub AppendMovieOptimised(ByVal title As String, ByVal year As String, ByVal duration As String, ByVal director As String, ByVal rating As String, ByVal takings As String)
        Dim fileName As String = "movies.csv"
        Dim newMovie As String = title & "," & year & "," & duration & "," & director & "," & rating & "," & takings

        Using writer As System.IO.StreamWriter = New System.IO.StreamWriter(fileName, True)
            writer.WriteLine(newMovie)
        End Using
    End Sub


End Module
